stdlib/public/core/BridgeObjectiveC.swift

```diff
@@ -640,3 +641,200 @@ public func _getObjCTypeEncoding<T>(_ type: T.Type) -> UnsafePointer<Int8> {
+  _: T.Type
+) -> T? {
+  var result: T?
+  T._conditionallyBridgeFromObjectiveC (x, result: &result)
+  return result
+}
+
+public // SPI(Foundation)
+protocol _NSSwiftValue: class {
+  init(_ value: Any)
+  var value: Any { get }
+  static var null: AnyObject { get }
+}
+
+@usableFromInline
+internal class _SwiftValue {
+  @usableFromInline
+  let value: Any
+  
+  @usableFromInline
+  init(_ value: Any) {
+    self.value = value
+  }
+  
+  @usableFromInline
+  static let null = _SwiftValue(Optional<Any>.none as Any)
+}
+
+/// COMPILER_INTRISIC
+@_silgen_name("swift_unboxFromSwiftValueWithType")
+public func swift_unboxFromSwiftValueWithType<T>(
+  _ source: inout AnyObject,
+  _ result: UnsafeMutablePointer<T>
+  ) -> Bool {
+
+  if source === _nullPlaceholder {
+    if let unpacked = Optional<Any>.none as? T {
+      result.initialize(to: unpacked)
+      return true
+    }
+  }
+    
+  if let box = source as? _SwiftValue {
+    if let value = box.value as? T {
+      result.initialize(to: value)
+      return true
+    }
+  } else if let box = source as? _NSSwiftValue {
+    if let value = box.value as? T {
+      result.initialize(to: value)
+      return true
+    }
+  }
+  
+  return false
+}
+
+@_silgen_name("_swift_extractDynamicValue")
+public func _extractDynamicValue<T>(_ value: T) -> AnyObject?
+
+@_silgen_name("_swift_bridgeToObjectiveCUsingProtocolIfPossible")
+public func _bridgeToObjectiveCUsingProtocolIfPossible<T>(_ value: T) -> AnyObject?
+
+@usableFromInline
+protocol _Unwrappable {
+  func unwrap() -> Any?
+}
+
+extension Optional: _Unwrappable {
+  func unwrap() -> Any? {
+    return self
+  }
+}
+
+// This is a best-effort tripmine for detecting the situation
+// (which should never happen) of Swift._SwiftValue and
+// Foundation._SwiftValue/Foundation.NSNull being used 
+// in the same process.
+
+@usableFromInline
+internal enum _SwiftValueFlavor: Equatable {
+  case stdlib
+  case foundation
+}
+
+@usableFromInline
+func _currentSwiftValueFlavor() -> _SwiftValueFlavor {
+  if _typeByName("Foundation._SwiftValue") as? _NSSwiftValue.Type != nil {
+    return .foundation
+  } else {
+    return .stdlib
+  }
+}
+
+@usableFromInline
+internal var _selectedSwiftValueFlavor: _SwiftValueFlavor = _currentSwiftValueFlavor()
+
+@usableFromInline
+internal func _assertSwiftValueFlavorIsConsistent() {
+  assert(_selectedSwiftValueFlavor == _currentSwiftValueFlavor())
+}
+
+@usableFromInline
+internal var _nullPlaceholder: AnyObject {
+  _assertSwiftValueFlavorIsConsistent()
+  if let foundationType = _typeByName("Foundation._SwiftValue") as? _NSSwiftValue.Type {
+    return foundationType.null
+  } else {
+    return _SwiftValue.null
+  }
+}
+
+@usableFromInline
+func _makeSwiftValue(_ value: Any) -> AnyObject {
+  _assertSwiftValueFlavorIsConsistent()
+  if let foundationType = _typeByName("Foundation._SwiftValue") as? _NSSwiftValue.Type {
+    return foundationType.init(value)
+  } else {
+    return _SwiftValue(value)
+  }
+}
```

Comment:
Is there a plan to retire `Foundation._SwiftValue` once this is all done?

Reply:
The problem is that _SwiftValue is an NSObject subclass on Apple platforms, and we probably can't undo that.

Reply:
How does that become important in practice?

Reply:
It becomes important when it interacts with CF API. Right now, calling CFCopy() will go through the NSCopying conformance that all SwiftValues implicitly have to, and that is still possible on Linux as well (you can toll-free bridge swift-corelibs-foundation’s NSArray/Dictionary to CFArray/DictionaryRef the same way you can on Darwin, and these may contain _SwiftValues and be set up to copy.)

Reply:
(It is also secondarily important for two reasons:
- It allows Linux to respond to `is/as? NSObject/NSCopying` the same way Darwin does, which can be a subtle issue in codebases where that question is posed in code that doesn’t know it may be receiving a boxed value;
- Existing boxing code in s-c-f already assumes that boxed values can be freely casted to NSObject (and uses its Hashable conformance to eg use them as dictionary keys); the churn was pretty big and would have introduced a number of forced casts and uses of AnyHashable, so I went instead with the solution that also he the property of being the closest match to Darwin.

Reply:
At a high level, I don't think changing the boxing behavior dynamically based on the presence of Foundation is workable—what happens when a program that doesn't use Foundation dynamically loads a library that does? `as? NSObject` working is not something we want to guarantee works with plain boxed values on Darwin. We control the CF implementation corelibs uses on Linux, so could we make CFCopy do the right thing with Swift boxes?

If we really must make boxed values be NSObject subclasses when Foundation is loaded, we should look into doing so in a way that doesn't invalidate boxed values that exist before Foundation is loaded. For instance, we could make it so that boxing uses heap objects that start out with non-class heap metadata, and overwrite the metadata with an NSObject subclass when Foundation gets loaded.